# Pair Programming - ANOVA

Hasta ahora habéis estado evaluando las características de vuestro set de datos y habéis hecho una gran exploración, es el momento de hacer vuestro primer ANOVA! En el ejercicio de hoy tendréis que hacer un ANOVA con vuestro datos y hacer una interpretación de los resultados.

NOTA Puede que vuestros datos no se ajusten o no cumplan todas las asunciones, no pasa nada, haced el ANOVA e interpretad los resultados. En próximas lecciones aprenderemos que podemos hacer cuando nos encontramos en esta situación.

In [9]:
import numpy as np
import pandas as pd
import random 

import matplotlib.pyplot as plt
import seaborn as sns

import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.multivariate.manova import MANOVA
from sklearn.preprocessing import StandardScaler

import statsmodels.api as sm
from statsmodels.formula.api import ols

plt.rcParams["figure.figsize"] = (10,8) 

In [10]:
df = pd.read_csv("data/adult.numericas_robust.csv", index_col = 0)

In [11]:
df.head(2)

work_class   education  education_yrs       marital_status  \
39                                                                      
50   Self-emp-not-inc   Bachelors      -0.333333   Married-civ-spouse   
38            Private     HS-grad       0.666667             Divorced   

            occupation    relationship ethnicity gender  capital_gain  \
39                                                                      
50     Exec-managerial         Husband     White   Male           0.0   
38   Handlers-cleaners   Not-in-family     White   Male           0.0   

    capital_lost  hours_week         country  salary  
39                                                    
50           0.0        -2.0   United-States   24006  
38           0.0         0.0   United-States   44640

In [12]:
anova = ols('salary ~  work_class  + education + education_yrs + marital_status + occupation + relationship + ethnicity + gender + capital_gain + capital_lost + hours_week + country' , data = df).fit()
sm.stats.anova_lm(anova)

df        sum_sq       mean_sq           F         PR(>F)
work_class          8.0  3.872754e+12  4.840942e+11  129.361393  1.285960e-214
education          15.0  1.432693e+13  9.551286e+11  255.232866   0.000000e+00
marital_status      6.0  1.873749e+13  3.122915e+12  834.516606   0.000000e+00
occupation         14.0  1.848825e+12  1.320589e+11   35.289255   9.945621e-96
relationship        5.0  2.311637e+11  4.623274e+10   12.354478   5.367586e-12
ethnicity           4.0  6.761329e+10  1.690332e+10    4.516966   1.199383e-03
gender              1.0  3.719241e+11  3.719241e+11   99.386898   2.244144e-23
country            41.0  2.820830e+11  6.880072e+09    1.838518   8.612880e-04
education_yrs       1.0  1.476838e+10  1.476838e+10    3.946460   4.697847e-02
capital_gain        1.0  1.360261e+10  1.360261e+10    3.634938   5.658752e-02
capital_lost        1.0  7.257960e+09  7.257960e+09    1.939498   1.637326e-01
hours_week          1.0  3.940715e+10  3.940715e+10   10.530520   1.175380e-03
Residual        32462.0  1.214788e+14  3.742185e+09         NaN            NaN

**Interpretación de la ANOVA:** REINTERPRETAR: SALARY

Es importante recalcar previo a la interpretación que los datos con los que trabajamos no son normales.
- **df:** Las variables categóricas tienen más rango de libertad, mientras que las numéricas sólo uno. Esto es porque las numéricas sólo tienen el rango de los números naturales.
- **sum_sq:** Esta es la variación atribuida al error. Podemos observar que, en este caso, es mínima.
- **mean_sq:** Este es el producto de la dos columnas anteriores.
- **F:** ANOVA supone que las medias de todos los grupos son iguales, calcula cuánto se desvían las medias iguales de la suposición. Una puntuación mayor significa que hay mayor diferencia entre las medias, por lo tanto, las que tienen valores más próximos entre sí van a tener medias más similares.
- **PR(>F):** Aquí podemos observar que el p-valor es en todos los casos inferior a 0,05, con lo cual, debemos rechazar la hipótesis nula y concluimos que estas variables sí influyen en nuestra variable predictora.

In [13]:
anova.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 salary   R-squared:                       0.247
Model:                            OLS   Adj. R-squared:                  0.245
Method:                 Least Squares   F-statistic:                     109.7
Date:                Fri, 27 Jan 2023   Prob (F-statistic):               0.00
Time:                        18:22:49   Log-Likelihood:            -4.0501e+05
No. Observations:               32560   AIC:                         8.102e+05
Df Residuals:                   32462   BIC:                         8.110e+05
Df Model:                          97                                         
Covariance Type:            nonrobust                                         
============================================================================================================
                                               coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------
Intercept                                 3.219e+04   6637.714      4.850      0.000    1.92e+04    4.52e+04
work_class[T. Federal-gov]                1.413e+04   2437.678      5.796      0.000    9351.054    1.89e+04
work_class[T. Local-gov]                  2091.2880   2096.454      0.998      0.319   -2017.839    6200.415
work_class[T. Never-worked]               2576.2835   2.32e+04      0.111      0.912   -4.29e+04     4.8e+04
work_class[T. Private]                    5017.8458   1667.918      3.008      0.003    1748.665    8287.027
work_class[T. Self-emp-inc]                2.04e+04   2426.083      8.408      0.000    1.56e+04    2.52e+04
work_class[T. Self-emp-not-inc]           -945.2674   2021.470     -0.468      0.640   -4907.423    3016.888
work_class[T. State-gov]                 -2460.1111   2312.047     -1.064      0.287   -6991.809    2071.587
work_class[T. Without-pay]               -1.923e+04   1.57e+04     -1.228      0.220   -4.99e+04    1.15e+04
education[T. 11th]                        1396.8677   2690.035      0.519      0.604   -3875.700    6669.435
education[T. 12th]                        5361.4851   3569.143      1.502      0.133   -1634.168    1.24e+04
education[T. 1st-4th]                    -4701.7002   5298.486     -0.887      0.375   -1.51e+04    5683.528
education[T. 5th-6th]                    -4244.3248   4113.264     -1.032      0.302   -1.23e+04    3817.826
education[T. 7th-8th]                    -8916.5367   3153.854     -2.827      0.005   -1.51e+04   -2734.866
education[T. 9th]                        -4853.4698   3375.535     -1.438      0.150   -1.15e+04    1762.704
education[T. Assoc-acdm]                  1.431e+04   2784.242      5.140      0.000    8854.688    1.98e+04
education[T. Assoc-voc]                   1.599e+04   2627.953      6.085      0.000    1.08e+04    2.11e+04
education[T. Bachelors]                   2.915e+04   2268.381     12.850      0.000    2.47e+04    3.36e+04
education[T. Doctorate]                   6.457e+04   3804.112     16.974      0.000    5.71e+04     7.2e+04
education[T. HS-grad]                     4922.4349   2104.034      2.340      0.019     798.451    9046.419
education[T. Masters]                     4.534e+04   2640.469     17.170      0.000    4.02e+04    5.05e+04
education[T. Preschool]                  -2701.7732   8892.832     -0.304      0.761   -2.01e+04    1.47e+04
education[T. Prof-school]                 6.499e+04   3442.241     18.880      0.000    5.82e+04    7.17e+04
education[T. Some-college]                1.025e+04   2156.991      4.750      0.000    6018.893    1.45e+04
marital_status[T. Married-AF-spouse]      2.213e+04   1.34e+04      1.653      0.098   -4105.339    4.84e+04
marital_status[T. Married-civ-s

R-cuadrado = Variación explicada / variación total
El R-cuadrado siempre está entre 0 y 100%:
0% indica que el modelo no explica ninguna porción de la variabilidad de los datos de respuesta en torno a su media.
100% indica que el modelo explica toda la variabilidad de los datos de respuesta en torno a su media.

**Interpretación del summary**

- **R-square:** Siempre se encuentra entre el 0 y el 100%, el nuestro está en 0.247, por lo tanto, esto nos quiere decir que la variables presidctoras explican en un 24,7% nuestra variable respuesta.
- **coef:** Estos datos nos indican cuánto son afectadas las variables por las que se relacionan con ellas. Vemos que las afectaciones son mínimas, de muy poquitas milésimas, tanto en positivo como en negativo.
- **std err:** Esta columna nos está diciendo que tenemos un gran margen de error en la estimación del coeficiente, seguramente debido a la cantidad de ouliers.
- **t:** Es el resultado de dividir los dos valores anteriores. Con esto estamos intentando averiguar en qué medida la afectación de los datos entre sí influye en el margen de error, es decir, si las variables se influyen también cuando no tienen correlación.

No obstante, dado que nuestros datos ni son normales ni cumplen ninguna de las asunciones, es ejercicio no es necesario. Con lo cual, no nos extraña que las conclusiones que extraemos de él sean estas.